ledger/heavy/executor/integration/backuper_test.go

```diff
@@ -0,0 +1,510 @@
+	"github.com/insolar/insolar/insolar"
+	"github.com/insolar/insolar/insolar/jet"
+	"github.com/insolar/insolar/insolar/pulse"
+	"github.com/insolar/insolar/insolar/store"
+	"github.com/insolar/insolar/instrumentation/inslogger"
+	"github.com/insolar/insolar/ledger/heavy/executor"
+	"github.com/stretchr/testify/require"
+)
+
+type testKey struct {
+	id uint64
+}
+
+func (t *testKey) ID() []byte {
+	bs := make([]byte, 8)
+	binary.PutUvarint(bs, t.id)
+	return bs
+}
+
+func (t *testKey) Scope() store.Scope {
+	return store.ScopeJetDrop
+}
+
+func addLastBackupFile(t *testing.T, to string, lastBackupedVersion uint64) {
+	backupInfo := executor.LastBackupInfo{
+		LastBackupedVersion: lastBackupedVersion,
+	}
+	rawInfo, err := json.MarshalIndent(backupInfo, "", "    ")
+	require.NoError(t, err)
+
+	err = ioutil.WriteFile(to, rawInfo, 0600)
+	println("============ addLastBackupFile: ", to)
+	require.NoError(t, err)
+}
+
+func makeBackuperConfig(t *testing.T, prefix string, badgerDir string) configuration.Ledger {
+
+	cwd, err := os.Getwd()
+	if err != nil {
+		require.NoError(t, err)
+	}
+
+	lastBackupedVersionFile := badgerDir + "/last_version.json"
+	addLastBackupFile(t, lastBackupedVersionFile, 0)
+
+	tmpDir := "/tmp/BKP/"
+
+	cfg := configuration.Backup{
+		ConfirmFile:          "BACKUPED",
+		MetaInfoFile:         "META.json",
+		TargetDirectory:      tmpDir + "/TARGET/" + prefix,
+		TmpDirectory:         tmpDir + "/TMP",
+		DirNameTemplate:      "pulse-%d",
+		BackupWaitPeriod:     10,
+		BackupFile:           "incr.bkp",
+		Enabled:              true,
+		PostProcessBackupCmd: []string{"bash", "-c", cwd + "/post_process_backup.sh"},
+		LastBackupInfoFile:   lastBackupedVersionFile,
+	}
+
+	err = os.MkdirAll(cfg.TargetDirectory, 0777)
+	require.NoError(t, err)
+	err = os.MkdirAll(cfg.TmpDirectory, 0777)
+	require.NoError(t, err)
+
+	return configuration.Ledger{
+		Backup: cfg,
+		Storage: configuration.Storage{
+			DataDirectory: badgerDir,
+		},
+	}
+}
+
+func clearData(t *testing.T, cfg configuration.Ledger) {
+	err := os.RemoveAll(cfg.Backup.TargetDirectory)
+	require.NoError(t, err)
+	err = os.RemoveAll(cfg.Backup.TmpDirectory)
+	require.NoError(t, err)
+}
+
+func TestBackuper(t *testing.T) {
+
+	tmpdir, err := ioutil.TempDir("", "bdb-test-")
+	require.NoError(t, err)
+	defer os.RemoveAll(tmpdir)
+
+	cfg := makeBackuperConfig(t, t.Name(), tmpdir)
+	defer clearData(t, cfg)
+
+	db, err := store.NewBadgerDB(badger.DefaultOptions(tmpdir))
+	require.NoError(t, err)
+	defer db.Stop(context.Background())
+
+	bm, err := executor.NewBackupMaker(context.Background(), db, cfg, insolar.GenesisPulse.PulseNumber, db)
+	require.NoError(t, err)
+
+	savedKeys := make(map[store.Key]insolar.PulseNumber, 0)
+
+	var stopWriting uint32
+	sgWriteStopped := sync.WaitGroup{}
+	sgWriteStopped.Add(1)
+
+	testPulse := insolar.GenesisPulse.PulseNumber + insolar.PulseNumber(rand.Int()%20000+1)
+	// writing data to db
+	go func() {
+		for i := 0; i < 2000000; i++ {
```

Comment:
Why 2000000?

Reply:
use` for i := 0; ; i++` or `for atomic.LoadUint32(&stopWriting) == 0` (in second case increment` i` inside cycle)

Reply:
to limit amount of written keys!
no problem